### PR TITLE
Spectators can enter and view other player cities

### PR DESCRIPTION
**Pull request by alkorolyov**

### CODE REVIEW

**Comment by yairm210** [comment]
It shows '38 files changed', please merge the latest master into your branch

**Comment by alkorolyov** [comment]
Done. Should be ready for merge now.

**Comment by alkorolyov** [comment]
Seems to be ready for merge now

**Comment by alkorolyov** [comment]
Seems to be ready for merge

File: core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt [hunk 1]
```diff
@@ -334,8 +334,8 @@ class EmpireOverviewScreen(private val viewingPlayer:CivilizationInfo, defaultPa
 
         for(civ in relevantCivs.filter { playerKnows(it) && !it.isDefeated() })
             for(diplomacy in civ.diplomacy.values.
-                    filter { !it.otherCiv().isBarbarian() && !it.otherCiv().isCityState()
-                            && playerKnows(it.otherCiv()) && !it.otherCiv().isDefeated()}){
+                    filter { !it.otherCiv().isBarbarian() && !it.otherCiv().isCityState() && !it.otherCiv().isSpectator()
```
Comment (yairm210) [inline core/src/com/unciv/ui/overviewscreen/EmpireOverviewScreen.kt:337]: Could be simplified to usMajorCiv

File: core/src/com/unciv/ui/cityscreen/CityScreen.kt [hunk 2]
```diff
@@ -134,20 +135,24 @@ class CityScreen(internal val city: CityInfo): CameraStageBaseScreen() {
...
                 update()
             }
+            if (viewingCiv.isSpectator()) annexCityButton.disable()
             razeCityButtonHolder.add(annexCityButton).colspan(cityPickerTable.columns)
         } else if(!city.isBeingRazed) {
             val razeCityButton = "Raze city".toTextButton()
             razeCityButton.labelCell.pad(10f)
             razeCityButton.onClick { city.isBeingRazed=true; update() }
-            if(!UncivGame.Current.worldScreen.isPlayersTurn || city.isOriginalCapital)
+            if(!UncivGame.Current.worldScreen.isPlayersTurn
```
Comment (yairm210) [inline core/src/com/unciv/ui/cityscreen/CityScreen.kt:144]: There seem to be a lot of 'is players turn and is not spectator' checks, that should probably be put in its own val
Maybe call is canChangeState or something

Comment (alkorolyov) [reply core/src/com/unciv/ui/cityscreen/CityScreen.kt:144]: Great idea! Added it to worldScreen and it simplified a lot of those "can I add/enable state changing button" checks.

File: core/src/com/unciv/ui/cityscreen/SpecialistAllocationTable.kt [hunk 3]
```diff
@@ -22,9 +22,11 @@ class SpecialistAllocationTable(val cityScreen: CityScreen): Table(CameraStageBa
             val assignedSpecialists = cityInfo.population.specialists.get(stat).toInt()
             val maxSpecialists = cityInfo.population.getMaxSpecialists().get(stat).toInt()
 
-            add(getUnassignButton(assignedSpecialists,stat))
+            val viewingCiv = UncivGame.Current.worldScreen.viewingCiv
+
+            if (!viewingCiv.isSpectator()) add(getUnassignButton(assignedSpecialists,stat))
```
Comment (yairm210) [inline core/src/com/unciv/ui/cityscreen/SpecialistAllocationTable.kt:27]: Should also be canChangeState

**Comment by yairm210** [comment]
Looks really good!

File: core/src/com/unciv/ui/cityscreen/SpecialistAllocationTable.kt (hunk 3 above)
Comment (alkorolyov) [reply core/src/com/unciv/ui/cityscreen/SpecialistAllocationTable.kt]: Done!

**Pull request merged**